[Serializer] Serialize and deserialize from abstract classes

sroze opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ø
| License       | MIT
| Doc PR        | Not yet

This PR adds a feature in the Serializer: allow to serialize and de-serialize abstract classes. Such feature is especially useful when dealing with domain objects.

# Example

Let's take the example of the following objects:
- `CodeRepository` defines a set of properties like `name` and `url`
- `GitHubCodeRepository` and `BitBucketCodeRepository` extends from the abstract `CodeRepository` class and adds a few properties.
- `Project` has a relation with a `codeRepository`, which has a type `CodeRepository`. 

At the moment, the serializer can't serialize/deserialize correctly this `Project` object has it doesn't know how to deal with this `CodeRepository` abstract object.

This feature allows the serializer to deal with such situation. The `ObjectNormalizer` has now access to a `ClassDiscriminatorResolver` that knows, for a given abstract class:
- Is the "type" property it needs to read/write to uniquely identify each sub-class
- What's the name of the "type" for each sub-class mapping

# Usage without Framework Bundle

```php
$discriminatorResolver = new ClassDiscriminatorResolver();
$discriminatorResolver->addClassMapping(CodeRepository::class, new ClassDiscriminatorMapping('type', [
    'github' => GitHubCodeRepository::class,
    'bitbucket' => BitBucketCodeRepository::class,
]));

$serializer = new Serializer(array(new ObjectNormalizer(null, null, null, null, $discriminatorResolver)), array('json' => new JsonEncoder()));

$serialized = $serializer->serialize(new GitHubCodeRepository());
// {"type": "github"}

$repository = $serializer->unserialize($serialized, CodeRepository::class, 'json');
// GitHubCodeRepository

```

# Usage with the Framework Bundle

```yaml
framework:
    serializer:
        discriminator_class_mapping:
            App\CodeRepository:
                 type_property: type
                 mapping:
                    github: App\GitHubCodeRepository
                    bitbucket: App\BitBucketCodeRepository
```

# Usage with Annotations/XML/YAML

```php
use Symfony\Component\Serializer\Annotation\DiscriminatorMap;

/**
 * @DiscriminatorMap(typeProperty="type", mapping={
 *    "first"="Symfony\Component\Serializer\Tests\Fixtures\AbstractDummyFirstChild",
 *    "second"="Symfony\Component\Serializer\Tests\Fixtures\AbstractDummySecondChild"
 * })
 */
abstract class AbstractDummy
{
    public $foo;

    public function __construct($foo = null)
    {
        $this->foo = $foo;
    }
}
```

# TODO

- [x] Working as standalone
- [x] Working with the framework bundle
- [x] Tests on mapping classes

sroze committed:
Add class discriminator mapping to resolve abstract classes
Add the `discriminator_class_mapping` in the framework extension
Discriminator map is compatible with interfaces

dunglas commented on src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorMapping.php:
```diff
@@ -0,0 +1,71 @@
+    private $typeProperty;
+    private $typesMapping;
+
+    public function __construct(string $typeProperty, array $typesMapping = array())
```
Should be compatible with PHP 5.4 to target 3.4.

dunglas commented on src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorMapping.php:
```diff
@@ -0,0 +1,71 @@
+        return $this->typeProperty;
+    }
+
+    public function addTypeMapping(string $type, string $typeClass)
```
Is it possible to remove this method? In the PropertyInfo component we used immutable objects.

dunglas commented on src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorResolver.php:
```diff
@@ -0,0 +1,76 @@
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+final class ClassDiscriminatorResolver
```
It should implement an interface as it's not a value object (it has a behavior).

dunglas commented on src/Symfony/Component/Serializer/Normalizer/ObjectNormalizer.php:
```diff
@@ -30,11 +32,52 @@ class ObjectNormalizer extends AbstractObjectNormalizer
+    /**
+     * {@inheritdoc}
+     */
+    protected function instantiateObject(array &$data, $class, array &$context, \ReflectionClass $reflectionClass, $allowedAttributes, string $format = null)
```
Can't you move this in the parent class? It will make it easier to support for this feature in other normalizers.

dunglas commented:
Great, I like it a lot. Would fit perfectly to denormalize JSON-LD documents (`@type`).

sroze commented:
@dunglas updated based on your requests and added the framework bundle bridge.

sroze commented:
Can't get the bottom of this Travis issue for PHP 7+ 🤔

afurculita commented on src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorResolver.php:
```diff
@@ -0,0 +1,78 @@
+     */
+    public function addClassMapping($class, ClassDiscriminatorMapping $mapping)
+    {
+        if (isset($this->mapping[$class])) {
```
How about injecting the mappings through a constructor and making this verification at container compile time, not runtime?

afurculita commented on src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorResolver.php:
```diff
@@ -0,0 +1,78 @@
+     */
+    public function getMappingForClass($class)
+    {
+        if (isset($this->mapping[$class])) {
```
I think `array_key_exists` can safely replace `isset` here

theofidry commented:
@sroze could you provide a sample of the problematic code? I do have similar relationships in my codebase and it works.

sroze commented:
@theofidry the tricky bit is the de-serialization, it would complain about not being able to instanciate an abstract class 😉

jderusse commented:
Is this feature extensible to everything? and not only abstract classes.

Would love to use `$message = $serializer->unserialize($data, MessageInterface::class, 'json');`
Without need to extend an abstract class nor writing a the list of existing mapping.

Mapping could be done by using the basename of the Entity classname (serialize), and reverse mapping by browsing a list of known namespaces and chaking if the class name exists.

WDYT?

ogizanagi commented:
I didn't look at this deep, but:
1. why do we expose both `framework.serializer.discriminator_class_mapping` and yml/xml/annotations for this?
2. why allowing to pass a `ClassDiscriminatorResolverInterface` to `ObjectNormalizer` instead of relying on the metadata already registered?

I agree this feature could be useful without abstract classes too.

nicolas-grekas commented on src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorMapping.php:
```diff
@@ -0,0 +1,77 @@
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+final class ClassDiscriminatorMapping
```
we don't make classes "final" by default, eg for data objects. Is it the case here? If there is no special reason, I'd suggest to remove them.

sroze replied:
Yes, this is a data object. Happy to remove the `final` even tho I believe closed-first is the best approach 😜

sroze replied:
@nicolas-grekas removed the final 👍

sroze commented:
> Would love to use $message = $serializer->unserialize($data, MessageInterface::class, 'json');

@jderusse yes, good point, I'll do this.

> Mapping could be done by using the basename of the Entity classname (serialize), and reverse mapping by browsing a list of known namespaces and chaking if the class name exists.

That's a very good point as well, this could potentially be automated using a `class name => class name` mapping. But I believe it would be better to start with an explicit configuration as I wouldn't personally advise to use class names within serialised documents. If we see people having no problem with this and see the DX can be drastically improved this way, why not :)

> why do we expose both framework.serializer.discriminator_class_mapping and yml/xml/annotations for this?
> why allowing to pass a ClassDiscriminatorResolverInterface to ObjectNormalizer instead of relying on the metadata already registered?

@ogizanagi The reason is that I started by exposing the `discriminator_class_mapping` service but @dunglas told me he would prefer to have some configuration looking like the existing one. I have kept both as I believe it gives us a greater configuration flexibility but happy to keep only one.

> What about allowing to call a resolver instead of a property?

@ogizanagi I don't get the question.

afurculita commented on src/Symfony/Component/Serializer/Annotation/DiscriminatorMap.php:
```diff
@@ -0,0 +1,70 @@
+     */
+    public function __construct(array $data)
+    {
+        if (!isset($data['typeProperty']) || !$data['typeProperty']) {
```
`array_key_exist` should be used instead of `isset` here

sroze replied:
Yes, `null` or a falsy contents should throw an exception, hence `isset`.

afurculita commented on src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorFromClassMetadata.php:
```diff
@@ -0,0 +1,84 @@
+    private $classMetadataFactory;
+
+    /**
+     * @param ClassMetadataFactoryInterface $classMetadataFactory
```
No need for this phpdoc

dunglas commented on src/Symfony/Component/Serializer/Annotation/DiscriminatorMap.php:
```diff
@@ -0,0 +1,70 @@
+            throw new InvalidArgumentException(sprintf('Parameter "typeProperty" of annotation "%s" cannot be empty.', get_class($this)));
+        }
+
+        if (!isset($data['mapping']) || empty($data['mapping'])) {
```
you can change `empty` to `!` like in the previous test (or just use `empty` and remove the call to `isset`).

dunglas commented on src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorFromClassMetadata.php:
```diff
@@ -0,0 +1,84 @@
+            $metadata = $this->classMetadataFactory->getMetadataFor($object);
+
+            if (null !== $metadata->getClassDiscriminatorMapping()) {
+                return $metadata->getClassDiscriminatorMapping();
```
Storing the result in a local cache will improve performance (especially in the case where reflection is involved).

sroze replied:
Good point, added local cache bellow 👍 (the class metadata factory is already cached)

dunglas commented on src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorFromClassMetadata.php:
```diff
@@ -0,0 +1,84 @@
+     */
+    public function getTypeForMappedObject($object)
+    {
+        if (null === ($mapping = $this->getMappingForMappedObject($object))) {
```
Parenthesis can be removed.

dunglas commented on src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php:
```diff
@@ -38,11 +41,17 @@
         parent::__construct($classMetadataFactory, $nameConverter);
 
         $this->propertyTypeExtractor = $propertyTypeExtractor;
+        $this->classDiscriminatorResolver = $classDiscriminatorResolver ?: (null !== $classMetadataFactory ? new ClassDiscriminatorFromClassMetadata($classMetadataFactory) : null);
```
We try to avoid imbricating ternaries (for readability). Can you switch to a `if` construct?

dunglas commented on src/Symfony/Component/Serializer/Normalizer/ObjectNormalizer.php:
```diff
@@ -98,6 +99,12 @@ protected function extractAttributes($object, $format = null, array $context = a
     protected function getAttributeValue($object, $attribute, $format = null, array $context = array())
     {
+        if ($this->classDiscriminatorResolver && $mapping = $this->classDiscriminatorResolver->getMappingForMappedObject($object)) {
+            if ($attribute == $mapping->getTypeProperty()) {
```
You can merge both `if` right?

dunglas commented on src/Symfony/Component/Serializer/Tests/Mapping/Loader/AnnotationLoaderTest.php:
```diff
@@ -52,6 +54,22 @@ public function testLoadGroups()
 
+    public function testLoadDiscriminatorMap()
+    {
+        $classMetadata = new ClassMetadata('Symfony\Component\Serializer\Tests\Fixtures\AbstractDummy');
```
You can use `AbstractDummy::class` here too (recommended for new code)

dunglas commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/php/serializer_discriminator_map.php:
```diff
@@ -0,0 +1,15 @@
+<?php
```
Those files should me removed now isn't it?

sroze commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/serializer.xml:
```diff
@@ -24,12 +24,18 @@
         <service id="serializer.property_accessor" alias="property_accessor" />
 
+        <!-- Discriminator Map -->
+        <service id="serializer.mapping.class_discriminator_resolver" class="Symfony\Component\Serializer\Mapping\ClassDiscriminatorFromClassMetadata">
```
It is private, as it's in the `defaults`. Added the alias.

sroze commented:
@jderusse this is now working with interfaces as well. 
@ogizanagi I've removed the configuration via the FrameworkBundle, only kept annotations/xml/yaml.
@dunglas added the resolver local cache and changed the CS.

sroze commented:
Test errors are normal as the functional tests are using `dev-master` components and therefore don't get the class introduced and used while creating the serializer. Should be ready 🤗

ogizanagi commented on src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php:
```diff
@@ -38,11 +41,21 @@
+     */
+    protected $classDiscriminatorResolver;
+
+    public function __construct(ClassMetadataFactoryInterface $classMetadataFactory = null, NameConverterInterface $nameConverter = null, PropertyTypeExtractorInterface $propertyTypeExtractor = null, ClassDiscriminatorResolverInterface $classDiscriminatorResolver = null)
```
Do we really want to keep exposing this `$classDiscriminatorResolver` argument? Configuring everything through metadata looks enough to me, isn't it?

dunglas replied:
I would keep it. We may leverage this extension point with API Platform to remove a lot of code related to `abstract` class handling. But we need to hook our own metadata system.

fabpot commented:
Moving this one to 4.1. There is no need to rush out such a new feature.

ostrolucky commented on src/Symfony/Component/Serializer/Annotation/DiscriminatorMap.php:
```diff
@@ -0,0 +1,70 @@
+     */
+    public function __construct(array $data)
+    {
+        if (!isset($data['typeProperty']) || !$data['typeProperty']) {
```
This is what's `empty` made for

sroze replied:
`empty` fails if the key do not exists.

ogizanagi replied:
@sroze : Hmm, no it doesn't 😅

dunglas replied:
And now that we target Symfony 4.1 (so PHP 7+), you can just use `??` :D

sroze replied:
fails = complains isn't it?

ogizanagi replied:
What I meant:

http://php.net/manual/en/function.empty.php
> empty() does not generate a warning if the variable does not exist.

jderusse replied:
from http://php.net/manual/en/function.empty.php

> No warning is generated if the variable does not exist. That means empty() is essentially the concise equivalent to !isset($var) || $var == false.

see https://3v4l.org/24O2e

sroze replied:
Used `empty` so we are all happy. Not sure about `??` within a `if` and throwing an exception...
```php
$this->typeProperty = $data['typeProperty'] ?? throw new ...;
```
??

sroze replied:
Oh. I was sure it was warninging, thanks ❤️

sroze replied:
Used it and added tests that proves it works 👍

sroze commented:
@fabpot changed the PR to be based on `master` and used PHP 7 things. The bot isn't happy about PHP7 yet tho 😉

afurculita commented on src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorFromClassMetadata.php:
```diff
@@ -0,0 +1,92 @@
+     * @var ClassMetadataFactoryInterface
+     */
+    private $classMetadataFactory;
+    private $mappingForMappedObjectCache = [];
```
I don't think the short syntax for arrays has been accepted. Php-CS-fixer still has the long syntax specified in the config file

sroze replied:
But if we target PHP7+ we need to accept this :)

afurculita replied:
I agree with you. But fabbot is still complaining. I think this rule should be removed from php_cs.dist and from fabbot, to permit new developments to use the short syntax. Would you like to open an issue / PR about this? At this point, according to this PR, they try to avoid merge conflicts.

sroze replied:
https://github.com/symfony/symfony/pull/24396 👍

afurculita commented on src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorMapping.php:
```diff
@@ -0,0 +1,71 @@
+     * @param string $typeProperty
+     * @param array  $typesMapping
+     */
+    public function __construct($typeProperty, array $typesMapping = [])
```
`string` can be added as typehint and the phpdoc removed

afurculita commented on src/Symfony/Component/Serializer/Mapping/ClassDiscriminatorResolverInterface.php:
```diff
@@ -0,0 +1,41 @@
+     *
+     * @return ClassDiscriminatorMapping|null
+     */
+    public function getMappingForClass(string $class);
```
A return type can be specified for each method

sroze replied:
We are targeting PHP 7.0, not 7.1, right? Can't add the return type when it can return `null` then 🤔

afurculita replied:
Symfony 4 supports PHP >= 7.1.3

sroze commented:
Updated with the `array()` syntax as discussed 👍

sroze commented:
Rebased. Tests are 💚 (except the obvious one with `DEPS=low`). 

/cc @dunglas @fabpot @ogizanagi

sroze commented:
Rebased on `master` & squashed all the commits. Failure on `DEPS=low` is expected until we merge it.

For you @symfony/deciders 👍

nicolas-grekas commented:
A failure on deps=low is almost never legitimate. Should be doubly checked.

sroze commented:
@nicolas-grekas a new class used by FrameworkBundle's configuration (in the declared services). Should we add it only when the class exists or something? 🤔

nicolas-grekas commented:
class_exists if possible is best, instead of bumping the minimum version

sroze committed:
Remove the class discriminator definition and alias if class do not exists

sroze commented:
Good shout, just made the [`FrameworkBundle` able to deal without the class](https://github.com/symfony/symfony/pull/24375/commits/72ee086f22fcbafac4fb52936b27a3cdc3dfddb6) 👍

sroze commented:
And it's green now. 💚

fabpot commented:
Thank you @sroze.

sroze commented:
I think we should do a blog post for this one. Javier, should we pair on
that? :)
On Thu, 7 Dec 2017 at 18:34, Fabien Potencier <notifications@github.com>
wrote:

> Closed #24375 <https://github.com/symfony/symfony/pull/24375>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/symfony/symfony/pull/24375#event-1377490172>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxHEXIVzy6LpTuTYyDEMUms0smEgFMpks5s-C_DgaJpZM4Pomri>
> .
>
